crates/ra_ide/src/completion/complete_unqualified_path.rs

```diff
@@ -4,20 +4,28 @@ use hir::ScopeDef;
 use test_utils::tested_by;
 
 use crate::completion::{CompletionContext, Completions};
+use hir::{Adt, ModuleDef};
 use ra_syntax::AstNode;
+use ra_syntax::SyntaxNode;
 
 pub(super) fn complete_unqualified_path(acc: &mut Completions, ctx: &CompletionContext) {
-    if !ctx.is_trivial_path {
-        return;
-    }
-
-    if ctx.is_pat_binding_or_const
+    if (!ctx.is_trivial_path && !ctx.is_pat_binding_or_const)
         || ctx.record_lit_syntax.is_some()
         || ctx.record_pat_syntax.is_some()
     {
         return;
     }
 
+    for ancestor in ctx.token.ancestors() {
```

Comment:
Hm, why do we need a loop here? I'd probably expect this to be just an `if let Some(ty) = ctx.expected_type_of(&node) {`, without the loop. Though, we might need a loop inside `expected_type_of`.

Reply:
Good point! I've moved the loop into `expected_type_of` now.